### PR TITLE
CSV report of Attendees

### PR DESCRIPTION
**Pull request by vidya-ram**
CSV report of Attendees with columns consisting of 'receipt no.', 'ticket type', 'attendee fullname', 'attendee email', 'attendee phone' and also attendee details like company, tshirt size as individual columns.

### CODE REVIEW

File: boxoffice/models/line_item.py [hunk 1]
```diff
@@ -189,6 +189,20 @@ def fetch_all_details(cls, item_collection):
...
 
     @classmethod
+    def fetch_assignee_details(cls, item_collection):
+        """
+        Returns invoice_no, ticket title, assignee fullname, assignee email, assignee phone and assignee details
+        for all the line items in a given item collection
+        """
+        line_item_join = db.outerjoin(cls, Assignee, db.and_(LineItem.id == Assignee.line_item_id,
+            Assignee.current == True)).join(Item).join(Order)
+        line_item_query = db.select([Order.invoice_no, Item.title, Assignee.fullname,
```
Comment (shreyas-satish) [inline boxoffice/models/line_item.py]: Useful to have `LineItem.line_item_seq` as well?

File: boxoffice/views/admin_report.py [hunk 2]
```diff
@@ -42,3 +41,28 @@ def row_handler(row):
...
+            if detail not in headers:
+                headers.append(detail)
+    rows = LineItem.fetch_assignee_details(item_collection)
+
+    def row_handler(row):
+        # Convert row to a dict
+        # row[-1] contains attendee details which is already a dict
+        dict_row = dict(zip(headers, row[:-1]))
+        # If attendee details is not empty add to the dict_row
+        if row[-1] is not None:
```
Comment (shreyas-satish) [inline boxoffice/views/admin_report.py]: Why would this be `None`? The query should eliminate this condition?

File: boxoffice/models/line_item.py [hunk 3]
```diff
@@ -189,6 +189,20 @@ def fetch_all_details(cls, item_collection):
         return db.session.execute(line_item_query).fetchall()
 
     @classmethod
+    def fetch_assignee_details(cls, item_collection):
+        """
+        Returns invoice_no, ticket title, assignee fullname, assignee email, assignee phone and assignee details
+        for all the line items in a given item collection
+        """
+        line_item_join = db.outerjoin(cls, Assignee, db.and_(LineItem.id == Assignee.line_item_id,
```
Comment (shreyas-satish) [inline boxoffice/models/line_item.py]: This should be a `join` and not an `outerjoin`. `join` will include all `line items` regardless of whether they have been assigned.

File: boxoffice/models/line_item.py (hunk 1 above)
Comment (vidya-ram) [inline boxoffice/models/line_item.py]: Added

File: boxoffice/views/admin_report.py (hunk 2 above)
Comment (vidya-ram) [inline boxoffice/views/admin_report.py]: Fixed

File: boxoffice/models/line_item.py (hunk 3 above)
Comment (vidya-ram) [inline boxoffice/models/line_item.py]: Fixed.

File: boxoffice/models/line_item.py [hunk 4]
```diff
@@ -189,6 +189,20 @@ def fetch_all_details(cls, item_collection):
...
+        """
+        Returns invoice_no, ticket title, assignee fullname, assignee email, assignee phone and assignee details
+        for all the line items in a given item collection
+        """
+        line_item_join = db.join(cls, Assignee, db.and_(LineItem.id == Assignee.line_item_id,
+            Assignee.current == True)).join(Item).join(Order)
+        line_item_query = db.select([Order.invoice_no, Item.title, LineItem.line_item_seq, Assignee.fullname,
+            Assignee.email, Assignee.phone, Assignee.details]).select_from(line_item_join).where(cls.status ==
+            LINE_ITEM_STATUS.CONFIRMED).where(Order.item_collection ==
+            item_collection).order_by('created_at')
```
Comment (shreyas-satish) [inline boxoffice/models/line_item.py]: Replace `created_at` with `ordered_at`.

File: boxoffice/views/admin_report.py [hunk 5]
```diff
@@ -42,3 +41,26 @@ def row_handler(row):
...
     return csv_response(headers, rows, row_handler=row_handler)
+
+
+@app.route('/admin/ic/<ic_id>/attendees.csv')
+@lastuser.requires_login
+@load_models(
+    (ItemCollection, {'id': 'ic_id'}, 'item_collection'),
+    permission='org_admin')
+def attendees_report(item_collection):
+    headers = ['receipt no.', 'ticket type', 'ticket no.', 'attendee fullname', 'attendee email', 'attendee phone']
```
Comment (shreyas-satish) [inline boxoffice/views/admin_report.py]: I suggest having `ticket no.` before `ticket type`.

File: boxoffice/models/line_item.py (hunk 4 above)
Comment (vidya-ram) [inline boxoffice/models/line_item.py]: Fixed

File: boxoffice/views/admin_report.py (hunk 5 above)
Comment (vidya-ram) [inline boxoffice/views/admin_report.py]: Fixed

File: boxoffice/models/line_item.py [hunk 6]
```diff
@@ -189,6 +189,20 @@ def fetch_all_details(cls, item_collection):
         return db.session.execute(line_item_query).fetchall()
 
     @classmethod
+    def fetch_assignee_details(cls, item_collection):
```
Comment (jace) [inline boxoffice/models/line_item.py]: This should be a regular method on the `ItemCollection` class. It does not belong on the `LineItem` class. Please move it. The `fetch_all_details` classmethod should move as well.

File: boxoffice/views/admin_report.py [hunk 7]
```diff
@@ -32,7 +31,7 @@ def admin_report(item_collection):
     (ItemCollection, {'id': 'ic_id'}, 'item_collection'),
     permission='org_admin')
 def tickets_report(item_collection):
-    headers = ['ticket id', 'order_id', 'invoice no', 'ticket type', 'base amount', 'discounted amount', 'final amount', 'discount policy', 'discount code', 'buyer fullname', 'buyer email', 'buyer phone', 'attendee fullname', 'attendee email', 'attendee phone', 'attendee details', 'utm_campaign', 'utm_source', 'utm_medium', 'utm_term', 'utm_content', 'utm_id', 'gclid', 'referrer', 'date']
+    headers = ['ticket id', 'order_id', 'receipt no.', 'ticket type', 'base amount', 'discounted amount', 'final amount', 'discount policy', 'discount code', 'buyer fullname', 'buyer email', 'buyer phone', 'attendee fullname', 'attendee email', 'attendee phone', 'attendee details', 'utm_campaign', 'utm_source', 'utm_medium', 'utm_term', 'utm_content', 'utm_id', 'gclid', 'referrer', 'date']
```
Comment (jace) [inline boxoffice/views/admin_report.py]: Are these headers meant to be identifier strings or English words? The `receipt no.` syntax seems inappropriate for both.

File: boxoffice/views/admin_report.py [hunk 8]
```diff
@@ -42,3 +41,26 @@ def row_handler(row):
...
+    rows = LineItem.fetch_assignee_details(item_collection)
+
+    def row_handler(row):
+        # Convert row to a dict
+        # row[-1] contains attendee details which is already a dict
+        dict_row = dict(zip(headers, row[:-1]))
+        dict_row.update(row[-1])
+        return dict_row
+
+    return csv_response(headers, rows, row_type='dict', row_handler=row_handler)
```
Comment (jace) [inline boxoffice/views/admin_report.py]: Return with mimetype `text/csv`

File: boxoffice/views/admin_report.py [hunk 9]
```diff
@@ -42,3 +41,26 @@ def row_handler(row):
...
+@load_models(
+    (ItemCollection, {'id': 'ic_id'}, 'item_collection'),
+    permission='org_admin')
+def attendees_report(item_collection):
+    headers = ['receipt no.', 'ticket no.', 'ticket id', 'ticket type', 'attendee fullname', 'attendee email', 'attendee phone']
+    for item in item_collection.items:
+        for detail in item.assignee_details.keys():
+            if detail not in headers:
+                headers.append(detail)
+    rows = LineItem.fetch_assignee_details(item_collection)
```
Comment (jace) [inline boxoffice/views/admin_report.py]: Update the method call to be on the `item_collection` instance.

File: boxoffice/models/line_item.py [hunk 10]
```diff
@@ -168,31 +168,52 @@ def is_cancellable(self):
...
+
+ItemCollection.fetch_all_details = property(fetch_all_details)
+
+
+def fetch_assignee_details(self):
+    """
+    Returns invoice_no, ticket title, assignee fullname, assignee email, assignee phone and assignee details
+    for all the line items in a given item collection
+    """
+    print "self", self
```
Comment (shreyas-satish) [inline boxoffice/models/line_item.py]: Remove `print`.

File: boxoffice/views/admin_report.py (hunk 8 above)
Comment (vidya-ram) [inline boxoffice/views/admin_report.py]: Added mimetype

File: boxoffice/models/line_item.py (hunk 6 above)
Comment (vidya-ram) [inline boxoffice/models/line_item.py]: Fixed

File: boxoffice/views/admin_report.py (hunk 7 above)
Comment (vidya-ram) [inline boxoffice/views/admin_report.py]: Does `Receipt no.` work?

**Pull request merged**